[Merged by Bors] - feat(Mathlib/Order/Ideal): ideals of sets are stable under directed unions and chains

samvang opened the pull request:
We prove that the directed union of a non-empty directed union of ideals of subsets of a preorder is an ideal, and in particular ideals of subsets of a preorder are stable under unions. In order to prove it, we also add `directed_on_sUnion`, a variant of `directed_on_iUnion`. 

---

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Ruben-VandeVelde commented on Mathlib/Order/Ideal.lean:
```diff
@@ -600,4 +601,22 @@ theorem cofinal_meets_idealOfCofinals (i : ι) : ∃ x : P, x ∈ 𝒟 i ∧ x 
+lemma isIdeal_sUnion_directed {C : Set (Set P)} (hidl : ∀ I ∈ C, IsIdeal I)
+    (hD : DirectedOn (· ⊆ ·) C) (hNe : C.Nonempty) : IsIdeal C.sUnion := by
+  refine ⟨isLowerSet_sUnion (fun I hI ↦ (hidl I hI).1), Set.nonempty_sUnion.2 ?_,
+  directed_on_sUnion hD (fun J hJ => (hidl J hJ).3)⟩
```
```suggestion
    directed_on_sUnion hD (fun J hJ => (hidl J hJ).3)⟩
```

Ruben-VandeVelde commented on Mathlib/Data/Set/Lattice.lean:
```diff
@@ -543,6 +543,14 @@ theorem directed_on_iUnion {r} {f : ι → Set α} (hd : Directed (· ⊆ ·) f)
     ⟨x, ⟨z, xf⟩, xa₁, xa₂⟩
 #align set.directed_on_Union Set.directed_on_iUnion
 
+theorem directed_on_sUnion {r} {S : Set (Set α)} (hd : DirectedOn (· ⊆ ·) S)
```
This should be `directedOn__sUnion`. The one above is wrong too - if it's not used too much, you might rename it in this PR, leaving a deprecated alias. Alternatively leave a TODO comment

samvang replied:
Done (and moved to line 2037), see comment below.

Ruben-VandeVelde commented on Mathlib/Data/Set/Lattice.lean:
```diff
@@ -543,6 +543,14 @@ theorem directed_on_iUnion {r} {f : ι → Set α} (hd : Directed (· ⊆ ·) f)
 
+theorem directed_on_sUnion {r} {S : Set (Set α)} (hd : DirectedOn (· ⊆ ·) S)
+    (h : ∀ x ∈ S, DirectedOn r x) : DirectedOn r (⋃₀ S) := by
+  simp only [DirectedOn, mem_sUnion, exists_imp]
```
Can this proof reuse the theorem above?

dagurtomas replied:
I would suggest doing the rename suggested above, and moving both lemmas to below `sUnion_eq_iUnion` and proving the latter as

```lean
theorem directedOn_sUnion {r} {S : Set (Set α)} (hd : DirectedOn (· ⊆ ·) S)
    (h : ∀ x ∈ S, DirectedOn r x) : DirectedOn r (⋃₀ S) := by
  rw [sUnion_eq_iUnion]
  exact directed_on_iUnion (directedOn_iff_directed.mp hd) (fun i ↦ h i.1 i.2)
```

samvang replied:
Thanks! I did what Dagur suggested.

Ruben-VandeVelde commented on Mathlib/Order/Ideal.lean:
```diff
@@ -600,4 +601,22 @@ theorem cofinal_meets_idealOfCofinals (i : ι) : ∃ x : P, x ∈ 𝒟 i ∧ x 
+variable [Preorder P]
+
+/-- A non-empty directed union of ideals of sets in a preorder is an ideal. -/
+lemma isIdeal_sUnion_directed {C : Set (Set P)} (hidl : ∀ I ∈ C, IsIdeal I)
```
I think`isIdeal_sUnion_of_directedOn`

Ruben-VandeVelde commented on Mathlib/Order/Ideal.lean:
```diff
@@ -600,4 +601,22 @@ theorem cofinal_meets_idealOfCofinals (i : ι) : ∃ x : P, x ∈ 𝒟 i ∧ x 
+
+/-- A union of a nonempty chain of ideals of sets is an ideal. -/
+lemma isIdeal_sUnion_nechain {C : Set (Set P)} (hidl : ∀ I ∈ C, IsIdeal I)
+    (hC : IsChain (· ⊆ ·) C) (hne : C.Nonempty) : IsIdeal C.sUnion :=
```
You use `hNe` before and `hne` here. It would be better to be consistent

Ruben-VandeVelde commented on Mathlib/Order/Ideal.lean:
```diff
@@ -600,4 +601,22 @@ theorem cofinal_meets_idealOfCofinals (i : ι) : ∃ x : P, x ∈ 𝒟 i ∧ x 
+  exact ⟨I, ⟨hI, (hidl I hI).2⟩⟩
+
+/-- A union of a nonempty chain of ideals of sets is an ideal. -/
+lemma isIdeal_sUnion_nechain {C : Set (Set P)} (hidl : ∀ I ∈ C, IsIdeal I)
```
`_of_isChain`

Ruben-VandeVelde commented:
And I should say: thanks for the contribution!

samvang commented:
Thanks very much for the help @Ruben-VandeVelde and @dagurtomas! I implemented your comments just now.

I also did "Resolve conversation" for each comment and changed back to "awaiting-review". I hope this is the correct workflow?

Ruben-VandeVelde commented on Mathlib/Data/Set/Lattice.lean:
```diff
@@ -2041,6 +2032,24 @@ theorem iUnion_univ_pi {ι : α → Type*} (t : (a : α) → ι a → Set (π a)
+    let ⟨z, zb₁, zb₂⟩ := hd b₁ b₂
+    let ⟨x, xf, xa₁, xa₂⟩ := h z a₁ (zb₁ fb₁) a₂ (zb₂ fb₂)
+    ⟨x, ⟨z, xf⟩, xa₁, xa₂⟩
+#align set.directed_on_Union Set.directedOn_iUnion
```
We want to give any downstream users of this theorem a pointer where it went, like this:

```suggestion
#align set.directed_on_Union Set.directedOn_iUnion

@[deprecated (since = "2024-05-05")]
alias directed_on_iUnion := directedOn_iUnion
```
(Writing on mobile, so I hope I got that right)

samvang replied:
```suggestion
#align set.directed_on_Union Set.directedOn_iUnion

@[deprecated (since := "2024-05-05")]
alias directed_on_iUnion := directedOn_iUnion
```

Ruben-VandeVelde reviewed (commented):
Yes, that's the correct workflow. Thanks!

samvang commented:
I just added the deprecation notice as per @Ruben-VandeVelde 's suggestion.

Ruben-VandeVelde reviewed (approved):
maintainer merge

jcommelin reviewed (approved):
Thanks :tada:

bors merge
